gpu_direct_rdma_access.c

```diff
@@ -838,6 +866,66 @@ struct rdma_device *rdma_open_device_server(struct sockaddr *addr)
     return NULL;
 }
 
+int rdma_reset_server_device(struct rdma_task_attr *attr)
+{
+	struct rdma_device *device = attr->local_buf_rdma->rdma_dev;
+	if (is_server(device)) {
+		fprintf(stderr, "Method \"rdma_reset_server_device()\" could be executed only by server side!\n");
+		return EOPNOTSUPP;
+	}
+	struct ibv_qp_attr      qp_attr;
+	enum ibv_qp_attr_mask   attr_mask;
+	memset(&qp_attr, 0, sizeof qp_attr);
+	
+	/* - - - - - - - Modify QP to ERR - - - - - - - */
+	qp_attr.qp_state = IBV_QPS_ERR;
+	attr_mask = IBV_QP_STATE;
+	DEBUG_LOG("ibv_modify_qp(qp = %p, qp_attr.qp_state = %d, attr_mask = 0x%x)\n",
+                      device->qp, qp_attr.qp_state, attr_mask);
+	if (ibv_modify_qp(device->qp, &qp_attr, attr_mask)) {
+		fprintf(stderr, "Failed to modify QP to ERR\n");
+		return 1;
+	}
+	
+	/* - - - - - - - FLUSH WORK COMPLETIONS - - - - - - - */
+	attr->flags |= RDMA_TASK_ATTR_ZERO_BYTE_MSG;
+	attr->wr_id = ZERO_BYTE_MSG_WR_ID;
+	rdma_submit_task(attr);
```

Comment:
maybe we can break the rdma_submit_task() into inner helper function which you call from the external rdma_submit_task() and from here.
this will allow easier passing of extra flags to do internal library logic

but exposing the internal lib flags to user API is just wrong.